lametro/views.py

```diff
@@ -811,3 +815,100 @@ def refresh_guid_trigger(request, refresh_key):
+
+
+class SmartLogicAPI(ListView):
+    api_key = SMART_LOGIC_KEY
+    query_format = 'https://cloud.smartlogic.com/svc/' \
+        + SMART_LOGIC_ENVIRONMENT + '/ses/CombinedModel/hints/{query}.json'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.token_refreshed = False
+
+    @property
+    def access_token(self):
+        session = self.request.session
+
+        if not session.get('smart_logic_token'):
+            session['smart_logic_token'] = self._generate_token()
+
+        return session['smart_logic_token']
+
+    def render_to_response(self, context):
+        '''
+        Return response as JSON.
+        '''
+        return JsonResponse(self.get_queryset(context))
+
+    def get_queryset(self, *args, **kwargs):
+        '''
+        Hit the SmartLogic endpoint. Tokens expire in two weeks, so if we get
+        an authentication-related status code, refresh the token and try again
+        once before failing out.
+        '''
+        url = self.query_format.format(query=self.request.GET['query'])
+        headers = {'Authorization': 'Bearer ' + self.access_token}
+
+        try:
+            response = requests.get(url, headers=headers)
+        except HTTPError as e:
+            print('Could not communicate with SmartLogic: {}'.format(e))
+        except Exception:
+            raise
+        else:
+            if response.status_code == 200:
+                return json.loads(response.content.decode('utf-8'))
+
+            elif response.status_code in (401, 403) and not self.token_refreshed:
+                self.request.session['smart_logic_token'] = self._generate_token()
+                self.token_refreshed = True
+                self.get_queryset(*args, **kwargs)
+
+            else:
+                print('Unexpected response from SmartLogic: {}'.format(response.content.decode('utf-8')))
+
+    def _generate_token(self):
+        '''
+        Get a JSON Web Token from the SmartLogic API.
+        '''
+        url = 'https://cloud.smartlogic.com/token'
+        params = {'grant_type': 'apikey', 'key': self.api_key}
+
+        try:
+            response = requests.post(url, data=params)
+        except HTTPError as e:
+            print('Could not authenticate with SmartLogic: {}'.format(e))
+            return None
+        except Exception:
+            raise
+        else:
+            data = json.loads(response.content.decode('utf-8'))
+            return data['access_token']
+
+
+def fetch_topic(request):
+    guid = request.GET['guid']
+
+    response = {}
+    response['guid'] = guid
+
+    try:
+        subject_guid = SubjectGuid.objects.get(guid=guid)
+        subject = subject_guid.name
+        response['subject_safe'] = urllib.parse.quote(subject)
+        response['status_code'] = 200
+    except MultipleObjectsReturned:
+        subjects = [s.name for s in SubjectGuid.objects.filter(guid=guid)]
+        subject = Subject.objects.get(subject__in=subjects)
```

Comment:
can you add a comment here for future us that we wrote this with the assumption that there should only ever be one of the two subjects associated with a duplicate subjectguid in use in the main subject table? if this is not the case, it will be nice to have that reminder. :-)